lib/fromZodIssue.ts

```diff
@@ -0,0 +1,113 @@
+import * as zod from 'zod';
+
+import {
+  ISSUE_SEPARATOR,
+  PREFIX,
+  PREFIX_SEPARATOR,
+  UNION_SEPARATOR,
+} from './config';
+import { joinPath } from './utils/joinPath';
+import { isNonEmptyArray } from './utils/NonEmptyArray';
+import { ValidationError } from './ValidationError';
+
+export type ZodIssue = zod.ZodIssue;
+
+export function getMessageFromZodIssue(props: {
+  issue: ZodIssue;
+  issueSeparator: string;
+  unionSeparator: string;
+  includePath: boolean;
+}): string {
+  const { issue, issueSeparator, unionSeparator, includePath } = props;
+
+  if (issue.code === 'invalid_union') {
+    return issue.unionErrors
+      .reduce<string[]>((acc, zodError) => {
+        const newIssues = zodError.issues
+          .map((issue) =>
+            getMessageFromZodIssue({
+              issue,
+              issueSeparator,
+              unionSeparator,
+              includePath,
+            })
+          )
+          .join(issueSeparator);
+
+        if (!acc.includes(newIssues)) {
+          acc.push(newIssues);
+        }
+
+        return acc;
+      }, [])
+      .join(unionSeparator);
+  }
+
+  if (includePath && isNonEmptyArray(issue.path)) {
+    // handle array indices
+    if (issue.path.length === 1) {
+      const identifier = issue.path[0];
+
+      if (typeof identifier === 'number') {
+        return `${issue.message} at index ${identifier}`;
+      }
+    }
+
+    return `${issue.message} at "${joinPath(issue.path)}"`;
+  }
+
+  return issue.message;
+}
+
+export function conditionallyPrefixMessage(
```

Comment:
[suggestion] Since this is used also by `fromZodError`, and what it accepts is not neither a `ZodIssue` nor a ` ZodError`, how about moving it on its own file as a reusable util?

Reply:
Sure